cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/cube/DefaultCube.java

```diff
@@ -439,30 +442,47 @@ private void incrementMetric(String metricName, long value) {
     return result;
   }
 
-  private Collection<TimeSeries> convertToQueryResult(CubeQuery query,
-                                                      Table<Map<String, String>, String,
-                                                        Map<Long, Long>> resultTable) {
-
-    List<TimeSeries> result = Lists.newArrayList();
+  private Collection<TimeSeries> convertToQueryResult(
+    CubeQuery query, Table<Map<String, String>, String, Map<Long, Long>> resultTable) {
+    List<TimeSeries> result = new ArrayList<>();
     // iterating each groupValue dimensions
     for (Map.Entry<Map<String, String>, Map<String, Map<Long, Long>>> row : resultTable.rowMap().entrySet()) {
       // iterating each measure
       for (Map.Entry<String, Map<Long, Long>> measureEntry : row.getValue().entrySet()) {
         // generating time series for a grouping and a measure
         int count = 0;
-        List<TimeValue> timeValues = Lists.newArrayList();
+        List<TimeValue> timeValues = new ArrayList<>();
         for (Map.Entry<Long, Long> timeValue : measureEntry.getValue().entrySet()) {
           timeValues.add(new TimeValue(timeValue.getKey(), timeValue.getValue()));
         }
         Collections.sort(timeValues);
-        PeekingIterator<TimeValue> timeValueItor = Iterators.peekingIterator(
-          new TimeSeriesInterpolator(timeValues, query.getInterpolator(), query.getResolution()).iterator());
-        List<TimeValue> resultTimeValues = Lists.newArrayList();
-        while (timeValueItor.hasNext()) {
-          TimeValue timeValue = timeValueItor.next();
-          resultTimeValues.add(new TimeValue(timeValue.getTimestamp(), timeValue.getValue()));
-          if (++count >= query.getLimit()) {
-            break;
+        List<TimeValue> resultTimeValues = new ArrayList<>();
+
+        MetricsAggregationOption aggregationOption = query.getAggregationOption();
+        // only do aggregation if the data points are larger than the required count
+        if (aggregationOption != null && aggregationOption.getCount() < timeValues.size()) {
+          int partitionSize = timeValues.size() / aggregationOption.getCount();
+          for (List<TimeValue> interval : Iterables.partition(timeValues, partitionSize)) {
+            if (aggregationOption.getAggregationOption().equals(MetricsAggregationOption.AggregationOption.LATEST)) {
+              resultTimeValues.add(interval.get(interval.size() - 1));
+            } else {
+              long sum = interval.stream().mapToLong(TimeValue::getValue).sum();
```

Comment:
here is where you need to apply the same aggregation that was used for the original query

Reply:
Here the aggregation is SUM, so we will need to add all the values up.